Firestore/core/src/local/leveldb_document_overlay_cache.cc

```diff
@@ -220,6 +266,43 @@ void LevelDbDocumentOverlayCache::ForEachOverlay(
   }
 }
 
+void LevelDbDocumentOverlayCache::DeleteLargestBatchIdIndexEntry(
+    absl::string_view encoded_key,
+    const LevelDbDocumentOverlayKey& decoded_key) {
+  db_->current_transaction()->Delete(
+      LevelDbDocumentOverlayLargestBatchIdIndexKey::Key(
+          user_id_, decoded_key.largest_batch_id(), encoded_key));
+}
+
+void LevelDbDocumentOverlayCache::AddLargestBatchIdIndexEntry(
+    absl::string_view encoded_key,
+    const LevelDbDocumentOverlayKey& decoded_key) {
+  db_->current_transaction()->Put(
+      LevelDbDocumentOverlayLargestBatchIdIndexKey::Key(
+          user_id_, decoded_key.largest_batch_id(), encoded_key),
+      "");
```

Comment:
I realize there are precedences that use empty values in "index" tables. But those are because the in those tables, the location is encoded as, for example a document key: https://osscs.corp.google.com/firebase-sdk/firebase-ios-sdk/+/master:Firestore/core/src/local/leveldb_key.h;l=635?q=leveldb_k&ss=firebase-sdk%2Ffirebase-ios-sdk

So the lookup process will be get the document key, then recalculate the primary key to read the actual document.

Here you could do the same, encode document key instead of a leveldb key, then do a recalculation; or slightly better performance-wise (but consumes more space): save the leveldb key as a value in the table, then no recalculation is required (it not being part of the key also helps with read/write performance).

Either way, the introduction of `LevelDbKey = 23,` in `leveldb_key` should not be needed.

Reply:
Done. I chose to use an empty value for the index entries and reconstruct the LevelDb key for the actual document at runtime. Since there will ultimately be 3 indexes (after my next PR) the savings in disk space could be more meaningful. It also removes a possibility of data getting out of sync somehow.

I also ended up making a bunch of other "improvements" to the code, things that were obvious now that I haven't looked at it for a few days. So the code churn since your last review is quite large relative to the code churn required just to implement the indexing improvement you suggested.

Here is a summary of the extra "improvements" I made:
- Improved the consistency of some variable and argument naming: a `DocumentKey` is now always `document_key` and any variable named `key` is referring to a LevelDb key. Previously, "key", "encoded_key", and "decoded_key" were used without much consistency, detracting from readability.
- Stopped passing around the "encoded" LevelDb keys; only the _decoded_ LevelDb keys are now passed around, simplifying method signatures.
- Removed the private overload of `GetOverlay()`, since it was never used.
- Removed a few one-line private methods that were only used in one place, and just inlined them (e.g. `DeleteLargestBatchIdIndexEntry()`).
- Added a constructor to `LevelDbDocumentOverlayKey` that initializes its instance variables to the given values; this enabled avoiding a few encode/decode round trips.
- Added an `Encode()` method to `LevelDbDocumentOverlayKey` to factor out boilerplate logic from other classes.
- Added methods to `LevelDbDocumentOverlayLargestBatchIdIndexKey` to convert to and from `LevelDbDocumentOverlayKey`, also to factor out boilerplate.
- Added unit test coverage for the new methods.